src/lpython/semantics/python_ast_to_asr.cpp

```diff
@@ -5962,6 +5997,108 @@ class BodyVisitor : public CommonVisitor<BodyVisitor> {
+                    (if seperator does not exist)  tuple:   (string, "", "")
+                    res_tuple_type: stores the type of each expression present in resulting 3-tuple
+                */
+                int seperator_pos = KMP_string_match(s_var, seperator);
+
+                Vec<ASR::expr_t *> res_tuple;
+                Vec<ASR::ttype_t *> res_tuple_type;
+                res_tuple.reserve(al, 3); 
+                res_tuple_type.reserve(al, 3);
+
+                std :: string first_res, second_res, third_res;
+
+                if(seperator_pos == -1) {
+                    /* seperator does not exist */
+                    first_res = s_var;
+                    second_res = "";
+                    third_res = "";
+                } else {
+                    first_res = s_var.substr(0, seperator_pos);
+                    second_res = seperator;
+                    third_res = s_var.substr(seperator_pos + seperator.size());
+                }
+
+                res_tuple.push_back(al, ASRUtils::EXPR(ASR::make_StringConstant_t(al, loc, s2c(al, first_res), arg_seperator_type)));
+                res_tuple.push_back(al, ASRUtils::EXPR(ASR::make_StringConstant_t(al, loc, s2c(al, second_res), arg_seperator_type)));
+                res_tuple.push_back(al, ASRUtils::EXPR(ASR::make_StringConstant_t(al, loc, s2c(al, third_res), arg_seperator_type)));
+
+                res_tuple_type.push_back(al, arg_seperator_type);
+                res_tuple_type.push_back(al,arg_seperator_type);
+                res_tuple_type.push_back(al,arg_seperator_type);
+
+                ASR::ttype_t *tuple_type = ASRUtils::TYPE(ASR::make_Tuple_t(al, loc, res_tuple_type.p, res_tuple_type.n));
+                tmp = ASR::make_TupleConstant_t(al, loc, res_tuple.p, res_tuple.size(), tuple_type);
+    
+            } else {
+                /*
+                    Invoked when Seperator argument is provided as a variable
+                    b: str = "ple"
+                    Eg: "apple".seperator(b)
+                */
+                ASR::symbol_t *fn_div = resolve_intrinsic_function(loc, "_lpython_str_partition");
+                Vec<ASR::call_arg_t> args;
+                args.reserve(al, 1);
+                ASR::call_arg_t str_arg;
+                str_arg.loc = loc;
+
+                ASR::ttype_t *str_type = ASRUtils::TYPE(ASR::make_Character_t(al, loc, 1, s_var.size(), nullptr, nullptr, 0));
+                str_arg.m_value = ASRUtils::EXPR(ASR::make_StringConstant_t(al, loc, s2c(al, s_var), str_type));
+
+                ASR::call_arg_t sub_arg;
+                sub_arg.loc = loc;
+                sub_arg.m_value = arg_seperator;
+                args.push_back(al, str_arg);
+                args.push_back(al, sub_arg);
+
+                tmp = make_call_helper(al, fn_div, current_scope, args, "_lpython_str_partition", loc);
+            }
+
+            return;
+
```

Comment:
Okay. Working on it @certik

Reply:
You mean like I should create a class Compile Time Evaluation and add create partition function to it ? @certik

Reply:
I made the changes. Can you please review if that is what you asked? @certik

Reply:
Do it like this:
```
ASR::expr_t *value = nullptr;
if (ASRUtils::expr_value(arg_seperator)) {
    value = string_compile_time_evaluation(s_var, arg_seperator, loc, arg_seperator_type, attr_name);
}
string_create_partition(loc, attr_name, s_var, arg_seperator, value);
```
As you can see, we construct `value` if there is a compile time value available, otherwise a nullptr. Then we always pass it to the `string_create_partition` function that creates the `FunctionCall` node and in there set the `value` of this node to the `value` argument.

Reply:
Yeah. I am doing that. But my only concern is that a typical `FunctionCall` Node is created using the following syntax: 

`make_FunctionCall_t(Allocator &al, const Location &a_loc, symbol_t* a_name, symbol_t* a_original_name, call_arg_t* a_args, size_t n_args, ttype_t* a_type, expr_t* a_value, expr_t* a_dt)`

For our usecase, let's say we go call `make_FunctionCall_t` from `create_partition` function. So, How can I go about getting the values for `FunctionCall` node creation - 

My string partition function looks as follows - 
```
void create_partition(const Location &loc, std::string &s_var, ASR::expr_t *arg_seperator,
        ASR::ttype_t *arg_seperator_type, ASR::expr_t* value)
```

Now let's try to fill in the values of `make_FunctionCall_t`.
1. `Allocator &al` - It is the global allocator `al` in our case (✔)
2. `const Location &a_loc` - It refers to `loc` which we are passing as argument in `create_partition() (✔)
3. `symbol_t* a_name` = We obtain it from below definition. No idea though whether it is correct or not. (❔)
```
ASR::symbol_t *fn_div = resolve_intrinsic_function(loc, "_lpython_str_partition");
```
4.  `symbol_t* a_original_name` = `nullptr` ✔
5. `call_arg_t* a_args`, = `args.p` (❔)
Now for this, I am creating a Vector of Call Arguments and getting their values as shown below
```
Vec<ASR::call_arg_t> args;
            args.reserve(al, 1);
            ASR::call_arg_t str_arg;
            str_arg.loc = loc;
            ASR::ttype_t *str_type = ASRUtils::TYPE(ASR::make_Character_t(al, loc, 1, s_var.size(), nullptr, nullptr, 0));
            str_arg.m_value = ASRUtils::EXPR(ASR::make_StringConstant_t(al, loc, s2c(al, s_var), str_type));
            ASR::call_arg_t sub_arg;
            sub_arg.loc = loc;
            sub_arg.m_value = arg_seperator;
            args.push_back(al, str_arg);
            args.push_back(al, sub_arg);
```
6. `size_t n_args` = `args.size()` ❔
7. `ttype_t* a_type`  = This should be the type of return variable, which I think should be a `tuple`. But no idea. (❌)
8. `expr_t* a_value` = `value` attribute obtained from args of `create_partition`✔
9. `expr_t* a_dt` = `nullptr`✔

@Thirumalai-Shaktivel @certik Can you check this once and help me fill in the blanks?

Reply:
I think args seems to be correct

We know that the Tuple type is a character, just create a new Tuple type (`tuple[str, str, str]`) and pass this as an argument.

Reply:
Created a `tuple_type` as follows - 
```
Vec<ASR::ttype_t *> res_tuple_type;
            res_tuple_type.reserve(al, 3);
            res_tuple_type.push_back(al, arg_seperator_type);
            res_tuple_type.push_back(al,arg_seperator_type);
            res_tuple_type.push_back(al,arg_seperator_type);
            ASR::ttype_t *tuple_type = ASRUtils::TYPE(ASR::make_Tuple_t(al, loc, res_tuple_type.p, res_tuple_type.n));
            tmp = ASR::make_FunctionCall_t(al, loc, fn_div, nullptr, 
                args.p, args.size(), tuple_type, nullptr, nullptr);
```

`arg_seperator_type` is a string. After all being done, I assigned the following to `tmp` - 
```
tmp = ASR::make_FunctionCall_t(al, loc, fn_div, nullptr, args.p, args.size(), tuple_type, value, nullptr);
```

It throws the following error - 
```
ASR verify pass error: ASR verify: FunctionCall::m_name must be a Function or Variable with FunctionType
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
  Binary file "/mnt/d/B.Tech/workspace/wsl-workspace/lpython/src/bin/lpython", in _start()
  Binary file "/lib/x86_64-linux-gnu/libc.so.6", in __libc_start_main()
  Binary file "/mnt/d/B.Tech/workspace/wsl-workspace/lpython/src/bin/lpython", in main()
  File "lpython.cpp", line 0, in (anonymous namespace)::compile_python_to_object_file(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, LCompilers::PassManager&, LCompilers::CompilerOptions&, bool, bool)
LCompilersException: Verify failed
```

I think there is some issue in `resolve_intrinsic_function`.